### PR TITLE
Batch pipeline tests

### PR DESCRIPTION
**Pull request by kamilsa**
### Description of the Change

Add tests checking batch pipeline. During its development bug with sending transactions to ordering gate one by one instead of batch has been fixed

### Benefits

Batches behaviour is tests

### Possible Drawbacks 

None
### Usage Examples or Tests

execute batch pipeline test

### CODE REVIEW

File: irohad/ordering/impl/ordering_service_impl.hpp [hunk 1]
```diff
@@ -133,6 +127,12 @@ namespace iroha {
 
       std::unique_ptr<shared_model::interface::ProposalFactory> factory_;
 
+      /**
```
Comment (muratovv) [inline irohad/ordering/impl/ordering_service_impl.hpp]: Why are these changes required?
I don't force you to revert it, I am just curious.

File: test/framework/batch_helper.hpp [hunk 2]
```diff
@@ -200,6 +200,13 @@ namespace framework {
 
       using HashesType = std::vector<shared_model::interface::types::HashType>;
 
+      namespace {
+        struct BatchMeta {
```
Comment (muratovv) [inline test/framework/batch_helper.hpp]: Maybe add some documentation here?

File: test/framework/batch_helper.hpp [hunk 3]
```diff
@@ -292,11 +297,33 @@ namespace framework {
...
+      return makeTestBatchTransactions(
+          shared_model::interface::types::BatchType::ATOMIC,
+          std::forward<TxBuilders>(builders)...);
+    }
+
+    /**
+     * Create test batch transactions from passed transaction builders with
+     * provided batch meta
+     * @tparam TxBuilders variadic types of tx builders
+     * @param batch_type type of the batch (either ATOMIC or ORDERED)
```
Comment (muratovv) [inline test/framework/batch_helper.hpp]: Better to remove 

> (either ATOMIC or ORDERED)

Because in future it may be possible other different options. Thus, this comment will mislead readers.

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  auto createFirstUser() {
+    return AcceptanceFixture::createUser(kFirstUser,
+                                         kFirstUserKeypair.publicKey())
+        .build()
+        .signAndAddSignature(kAdminKeypair)
+        .finish();
+  }
+
+  /**
+   * Create transaction to create second user
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: Seems, that both lines duplicate each other. Maybe remove one of them?

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+  /**
+   * Create transaction to create asset and add its given amount to given user
+   * @param account_id account for which amount of asset is added
+   * @param asset_name name of the asset to be created and added to the account
+   * @param amount amount of the asset to be added to the account
+   * @param keypair is used to sign transaction
+   * @return transaction with create asset and add asset quantity commands
+   */
+  auto createAndAddAssets(std::string account_id,
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: [optional]
I suggest reworking types with regard to shared_model types.

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+  /**
+   * Create builder for transaction to do transfer between given users with
+   * given amount of given asset
+   * @param src_account_id source accound id
+   * @param dest_account_id destination account id
+   * @param asset_name name of the asset (without domain) to be transferred
+   * @param amount amount of asset to be transferred
+   * @return transaction builder with transfer asset command
+   */
+  auto prepareTransferAssetBuilder(const std::string &src_account_id,
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: same as before

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        .createdTime(iroha::time::now())
+        .transferAsset(src_account_id,
+                       dest_account_id,
+                       asset_name + "#" + kDomain,
+                       "",
+                       amount);
+  }
+
+  /**
+   * Takes transaction and sign it wit provided signature
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: s/sing/signs
s/wit/with

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+      .sendTxSequenceAwait(
+          transaction_sequence,
+          [&transaction_sequence](const auto &block) {
+            // check that transactions from block are the same as transactions
+            // from transaction sequence
+            ASSERT_THAT(block->transactions(),
+                        Pointwise(RefAndPointerEq(),
+                                  transaction_sequence.transactions()));
+          })
+      .done();
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: Invokation of `done` is not required. Finalization will happen in dtor.

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+                ElementsAre(
+                    RefAndPointerEq(transaction_sequence.transactions()[0]),
+                    RefAndPointerEq(transaction_sequence.transactions()[2])));
+          })
+      .done();
+}
+
+INSTANTIATE_TEST_CASE_P(BatchPipelineParameterizedTest,
+                        BatchPipelineTest,
+                        // note additional comma is needed to make it compile
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: Nice explanation, thanks!

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+                        // https://github.com/google/googletest/issues/1419
+                        Values(interface::types::BatchType::ATOMIC,
+                               interface::types::BatchType::ORDERED), );
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: Regards to https://soramitsu.atlassian.net/browse/IR-1537 task. Also, there is required to write a test with MST and signatures.
I have missed this test or there are some troubles with that?

File: irohad/ordering/impl/ordering_service_impl.hpp (hunk 1 above)
Comment (kamilsa) [reply irohad/ordering/impl/ordering_service_impl.hpp]: it caused a warning with order of fields initialization

File: test/framework/batch_helper.hpp [hunk 11]
```diff
@@ -348,12 +338,24 @@ namespace framework {
 
       // makes clang avoid sending warning with unused function
       // (makeTxBatchCollection)
-      internal::makeTxBatchCollection(batch_meta);
+      //      internal::makeTxBatchCollection(batch_meta);
```
Comment (muratovv) [inline test/framework/batch_helper.hpp]: now this line looks misleading. Maybe remove it?

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 12]
```diff
@@ -56,9 +55,9 @@ class BatchPipelineTest
    * @param keypair is used to sign transaction
    * @return transaction with create asset and add asset quantity commands
    */
-  auto createAndAddAssets(std::string account_id,
-                          std::string asset_name,
-                          std::string amount,
+  auto createAndAddAssets(const interface::types::AccountIdType &account_id,
+                          const interface::types::AssetNameType &asset_name,
+                          const std::string &amount,
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: Why amount is not sh_m type?

File: test/integration/pipeline/batch_pipeline_test.cpp [hunk 13]
```diff
@@ -275,16 +272,13 @@ TEST_F(BatchPipelineTest, InvalidOrderedBatch) {
...
-                ElementsAre(
-                    RefAndPointerEq(transaction_sequence.transactions()[0]),
-                    RefAndPointerEq(transaction_sequence.transactions()[2])));
-          })
-      .done();
+      .sendTxSequenceAwait(transaction_sequence, [&](const auto block) {
+        ASSERT_THAT(
+            block->transactions(),
+            ElementsAre(
+                RefAndPointerEq(transaction_sequence.transactions()[0]),
```
Comment (muratovv) [inline test/integration/pipeline/batch_pipeline_test.cpp]: Please add to documentation fact that second transaction is stateful invalid.
